src/main/scala/tech/cryptonomic/conseil/tezos/NodeOperations.scala

```diff
@@ -0,0 +1,552 @@
+
+    val parseMichelsonScripts: Block => Block =
+      withLoggingContext[Block, String](makeContext = block => s"block with hash ${block.data.hash.value}") {
+        implicit ctx =>
+          val codeAlter = codeLens.modify(toMichelsonScript[MichelsonSchema, String])
+          val storageAlter = storageLens.modify(toMichelsonScript[MichelsonInstruction, String])
+          val parametersAlter = parametersLens.modify(toMichelsonScript[MichelsonInstruction, String])
+
+          codeAlter compose storageAlter compose parametersAlter
+      }
+
+    getBareBlock(hash, offset).flatMap(
+      data =>
+        (getAllOperationsAndAccountsForBlock(data), getCurrentVotesForBlock(data)).mapN {
+          case ((operations, accountIds), votes) =>
+            parseMichelsonScripts(Block(data, operations, votes)) -> accountIds
+        }
+    )
+  }
+
+  /**
+    * Fetches a single block from the chain without associated data
+    *
+    * @param hash   hash of the block
+    * @param offset pass it to get a relative predecessor of the reference hash
+    * @tparam Eff   the result effect
+    * @return       the block data
+    */
+  def getBareBlock[Eff[_]: MonadThrow: DataFetcher.Std[?[_], BlockCoordinates, BlockData]](
+    hash: BlockHash,
+    offset: Option[Offset] = None
+  ): Eff[BlockData] =
+    getFetcher.run(hash -> offset.getOrElse(0))
+
+  /**
+    * Gets the block head along with associated data.
+    *
+    * @tparam Eff the result effect
+    * @return     Block head
+    */
+  def getBlockHead[Eff[_]: MonadThrow](
+    implicit blockFetcher: DataFetcher.Std[Eff, BlockCoordinates, BlockData],
+    opsAndAccountsFetch: DataFetcher.Std[Eff, BlockHash, (List[OperationsGroup], List[AccountId])],
+    quorumFetch: DataFetcher.Std[Eff, BlockCoordinates, Option[Int]],
+    proposalFetch: DataFetcher.Std[Eff, BlockCoordinates, Option[ProtocolId]]
+  ): Eff[Block] =
+    getBlock(blockHeadHash).map(_._1)
+
+  /**
+    * Gets just the block head without associated data.
+    *
+    * @tparam Eff the result effect
+    * @return     Block head
+    */
+  def getBareBlockHead[Eff[_]: MonadThrow: DataFetcher.Std[?[_], BlockCoordinates, BlockData]]: Eff[BlockData] =
+    getBareBlock(blockHeadHash)
+
+  /**
+    * Fetches both operations for a block and accounts referenced by those operations
+    * @param block core data of the block
+    * @tparam Eff  the result effect
+    * @return      The list of operations and account ids
+    */
+  def getAllOperationsAndAccountsForBlock[Eff[_]: MonadThrow: DataFetcher.Std[
+    ?[_],
+    BlockHash,
+    (List[OperationsGroup], List[AccountId])
+  ]](
+    block: BlockData
+  ): Eff[(List[OperationsGroup], List[AccountId])] =
+    //This is a workaround for the Tezos node returning a 404 error when asked for the operations or accounts of the genesis blog, which seems like a bug.
+    if (isGenesis(block))
+      (List.empty[OperationsGroup], List.empty[AccountId]).pure[Eff]
```

Comment:
Is there an Applicative available in the implicit scope here?

Reply:
The `Eff[_]: MonadThrow]` constraint means there must be an 
`implicit mt: MonadThrow[Eff]`

which is, as the name imply, a `Monad` that additionally allows to raise and catch errors.

Every Monad is by definition also an Applicative Functor, always.

Reply:
There is a hierarchy in operations available in the following type classes:
Functor < Applicative < Monad

Meaning that every Applicative is a Functor and every Monad is Applicative (hence Functor).

That's because those higher in this ordering have more powerful operations in terms of which you can define those lower in the hierarchy, consistently with the higher operations.

E.g. you can define easily `map` (for Functor) in terms of `flatMap and pure` (available in Monad)